src/stream/successor.rs

```diff
@@ -0,0 +1,98 @@
+use std::marker::PhantomData;
+use std::pin::Pin;
+
+use crate::future::Future;
+use crate::stream::Stream;
+use crate::task::{Context, Poll};
+
+/// A stream that yields elements by calling an async closure with the previous value as an
+/// argument
+///
+/// This stream is constructed by [`successor`] function
+///
+/// [`successor`]: fn.successor.html
+#[derive(Debug)]
+pub struct Successor<F, Fut, T>
+where
+    Fut: Future<Output = T>,
+{
+    successor: F,
+    future: Option<Fut>,
+    next: T,
+    _marker: PhantomData<Fut>,
+}
+
+/// Creates a new stream where to produce each new element a clousre is called with the previous
+/// value.
+///
+/// #Examples
+///
+/// ```
+/// # fn main() { async_std::task::block_on(async {
+/// #
+/// use async_std::prelude::*;
+/// use async_std::stream;
+///
+/// let s = stream::successor(22, |val| {
+///     async move {
+///         val + 1
+///     }
+/// });
+///
+/// pin_utils::pin_mut!(s);
+/// assert_eq!(s.next().await, Some(23));
+/// assert_eq!(s.next().await, Some(24));
+/// assert_eq!(s.next().await, Some(25));
+/// #
+/// # }) }
+///
+/// ```
+pub fn successor<F, Fut, T>(start: T, func: F) -> Successor<F, Fut, T>
+where
+    F: FnMut(T) -> Fut,
+    Fut: Future<Output = T>,
+    T: Copy,
```

Comment:
In order to be an async version of `successors` from the standard library, we'd need to:

1. Remove the `T: Copy` bound.
2. Change the bound on `F` to `FnMut(&T) -> Fut`.
3. Change the output type of the future to `Option<T>`.
4. Change the type of the `start` argument to `Option<T>`.

See here for the type signature in the standard library:
https://doc.rust-lang.org/std/iter/fn.successors.html

Reply:
I am not sure how to drop the `T: Copy` bounce, as I need to pass a reference to it into the passed in function _and_ return it in `Poll::Ready(value)`.
Unless we go all the way to Rc?

Reply:
I've done 3 & 4, but I am not sure how to tackle 1 & 2.

Reply:
@felipesere a way to do this is by always staying one behind, and only returning the result of the previous computation once the result of the next computation is known. Until `None` is hit, at which point we flush. If you look closely, this is how the iter version works too: https://doc.rust-lang.org/src/core/iter/sources.rs.html#611. `item` is only returned *after* next has been computed.